server/tests/1userSignup.test.js

```diff
@@ -0,0 +1,242 @@
+      firstName: 'John',
+      lastName: 'Doe',
+      email: 'johndoe@gmail.com',
+      username: 'Johnny',
+      password: 'Opcutuyh',
+      confirmPassword: 'Opcutuyh',
+    }).end((req, res) => {
+      res.status.should.eql(422);
+      res.body.should.be.a('object');
+      res.body.should.have.property('status').eql('error');
+      res.body.should.have.property('errors');
+      res.body.errors.should.be.a('object');
+      res.body.errors.should.have.property('password').include('Password most contain an Upper case letter, a lower case letter and a number.');
+      done();
+    });
+  });
+
+  it('should return 400 when a user does not provide a confirmPassword', (done) => {
+    chai.request(app).post('/api/auth/signup').send({
+      firstName: 'John',
+      lastName: 'Doe',
+      email: 'johndoe@gmail.com',
+      username: 'Johhny',
+      password: 'Opcut2way',
+    }).end((req, res) => {
+      res.status.should.eql(400);
+      res.body.should.be.a('object');
+      res.body.should.have.property('status').eql('error');
+      res.body.should.have.property('errors');
+      res.body.errors.should.be.a('object');
+      res.body.errors.should.have.property('confirmPassword').include('The confirmPassword field is required.');
+      done();
+    });
+  });
+
+  it('should return 400 when a user provides a confirmPassword that is not equl to the password', (done) => {
+    chai.request(app).post('/api/auth/signup').send({
+      firstName: 'John',
+      lastName: 'Doe',
+      email: 'johndoe@gmail.com',
+      username: 'Johhny',
+      password: 'Opcut2way',
+      confirmPassword: 'Opcut2wau',
+    }).end((req, res) => {
+      res.status.should.eql(400);
+      res.body.should.be.a('object');
+      res.body.should.have.property('status').eql('error');
+      res.body.should.have.property('errors');
+      res.body.errors.should.be.a('object');
+      res.body.errors.should.have.property('confirmPassword').include('The confirmPassword and password fields must match.');
+      done();
+    });
+  });
+
+  it('should return 409 when a user with the email already exists', (done) => {
+    chai.request(app).post('/api/auth/signup').send({
+      firstName: 'John',
+      lastName: 'Doe',
+      email: 'johndoe@gmail.com',
+      username: 'Johnny3',
+      password: 'Opcut2way',
+      confirmPassword: 'Opcut2way',
+    }).end((req, res) => {
+      res.status.should.eql(409);
+      res.body.should.be.a('object');
+      res.body.should.have.property('status').eql('fail');
+      res.body.should.have.property('message').eql('User with email: johndoe@gmail.com already exists.');
+      done();
+    });
+  });
+
+  it('should return 409 when a user with the username already exists', (done) => {
+    chai.request(app).post('/api/auth/signup').send({
+      firstName: 'John',
+      lastName: 'Doe',
+      email: 'johndoe1@gmail.com',
+      username: 'Johnny',
+      password: 'Opcut2way',
+      confirmPassword: 'Opcut2way',
```

Comment:
These are some excellent tests! Good completing the task and testing it extensively. We should try as much as possible to write DRY code. You may not be sending the same object to the '/api/auth/signup' route but they do have a lot in common. Is there a way to minimize this duplication?

Reply:
Do remember to update the .eslintrc.json to fix un-necessary hound comments.